src/vs/workbench/contrib/snippets/browser/surroundWithSnippet.ts

```diff
@@ -4,57 +4,136 @@
+import { EditorAction2, registerEditorContribution } from 'vs/editor/browser/editorExtensions';
 import { EditorContextKeys } from 'vs/editor/common/editorContextKeys';
 import { SnippetController2 } from 'vs/editor/contrib/snippet/browser/snippetController2';
 import { localize } from 'vs/nls';
 import { registerAction2 } from 'vs/platform/actions/common/actions';
 import { IClipboardService } from 'vs/platform/clipboard/common/clipboardService';
-import { ContextKeyExpr } from 'vs/platform/contextkey/common/contextkey';
+import { ContextKeyExpr, IContextKeyService } from 'vs/platform/contextkey/common/contextkey';
 import { IInstantiationService, ServicesAccessor } from 'vs/platform/instantiation/common/instantiation';
 import { pickSnippet } from 'vs/workbench/contrib/snippets/browser/snippetPicker';
-import { ISnippetsService } from 'vs/workbench/contrib/snippets/browser/snippets.contribution';
+import { ISnippetsService } from './snippets.contribution';
+import { CancellationToken } from 'vs/base/common/cancellation';
+import { Disposable } from 'vs/base/common/lifecycle';
+import { ITextModel } from 'vs/editor/common/model';
+import { CodeAction, CodeActionProvider, CodeActionContext, CodeActionList } from 'vs/editor/common/languages';
+import { CodeActionKind } from 'vs/editor/contrib/codeAction/browser/types';
+import { ILanguageFeaturesService } from 'vs/editor/common/services/languageFeatures';
+import { Range } from 'vs/editor/common/core/range';
+import { Selection } from 'vs/editor/common/core/selection';
+import { Snippet } from 'vs/workbench/contrib/snippets/browser/snippetsFile';
 
+const options = {
+	id: 'editor.action.surroundWithSnippet',
+	title: {
+		value: localize('label', 'Surround With Snippet...'),
+		original: 'Surround With Snippet...'
+	},
+	precondition: ContextKeyExpr.and(
+		EditorContextKeys.writable,
+		EditorContextKeys.hasNonEmptySelection
+	),
+	f1: true,
+};
 
-registerAction2(class SurroundWithAction extends EditorAction2 {
+class SurroundWithSnippet {
+	constructor(
+		private readonly _editor: ICodeEditor,
+		@ISnippetsService private readonly _snippetService: ISnippetsService,
+		@IClipboardService private readonly _clipboardService: IClipboardService,
+		@IContextKeyService private readonly _contextKeyService: IContextKeyService,
+		@IInstantiationService private readonly _instaService: IInstantiationService,
+	) { }
 
-	constructor() {
-		super({
-			id: 'editor.action.surroundWithSnippet',
-			title: { value: localize('label', 'Surround With Snippet...'), original: 'Surround With Snippet...' },
-			precondition: ContextKeyExpr.and(EditorContextKeys.writable, EditorContextKeys.hasNonEmptySelection),
-			f1: true
-		});
-	}
+	async getSurroundableSnippets(): Promise<Snippet[]> {
+		if (!this._editor.hasModel()) {
+			return [];
+		}
 
-	async runEditorCommand(accessor: ServicesAccessor, editor: ICodeEditor, ...args: any[]) {
+		const model = this._editor.getModel();
```

Comment:
Only extract this into a seprate function, e.g `function surroundateSnippet(accessor, model)` and leave the action as-is otherwise

Reply:
I didn't fully understand this commment. So, I took this as a hint to refactor from class-based to function-based implementation. So, I extracted the functions out and  added an `accessor: ServicesAccessor` parameter to them to help resolve their dependencies.